sklearn/inspection/_partial_dependence.py

```diff
@@ -79,30 +84,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
 
     values = []
     for feature in range(X.shape[1]):
-        uniques = np.unique(_safe_indexing(X, feature, axis=1))
-        if uniques.shape[0] < grid_resolution:
-            # feature has low resolution use unique vals
-            axis = uniques
-        else:
-            # create axis based on percentiles and grid resolution
-            emp_percentiles = mquantiles(
-                _safe_indexing(X, feature, axis=1), prob=percentiles, axis=0
-            )
-            if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+        if feature in custom_values:
+            # Use values in the custom range
+            feature_range = custom_values[feature]
+            if not isinstance(feature_range, np.ndarray):
+                feature_range = np.array(feature_range)
+            if feature_range.ndim != 1:
                 raise ValueError(
-                    "percentiles are too close to each other, "
-                    "unable to build the grid. Please choose percentiles "
-                    "that are further apart."
+                    "Grid for feature {} is not a one-dimensional array. Got {}"
+                    " dimensions".format(feature, feature_range.ndim)
+                )
+            axis = feature_range
```

Comment:
I'll prefer to error check before hand the loop since checking `custom_values` is more of an input validation: (It also reduces the code that is indented:

```python
custom_values = {k: np.asarray(v) for k, v in custom_values.items()}
if any(v.ndim != 1 for v in custom_values.values()):
    raise ValueError(...)

for feature in range(X.shape[1]):
    if feature in custom_values:
        axis = custom_values[feature]
    else:
        ...
```

In the erorr message, I think it's enough to say that `custom_values` contain non-1d data. The feature index does not correspond to the location of the original `X` anymore so it can be confusing.

Also, we can add a new test to `test_grid_from_X_error` to check the `custom_values` validation